packages/jupyter-ai-magics/jupyter_ai_magics/__init__.py

```diff
@@ -1,34 +1,72 @@
+from langchain_core._import_utils import import_attr
+
 from ._version import __version__
 
-# expose embedding model providers on the package root
-from .embedding_providers import (
-    BaseEmbeddingsProvider,
-    GPT4AllEmbeddingsProvider,
-    HfHubEmbeddingsProvider,
-    QianfanEmbeddingsEndpointProvider,
-)
-from .exception import store_exception
-from .magics import AiMagics
-
-# expose JupyternautPersona on the package root
-# required by `jupyter-ai`.
-from .models.persona import JupyternautPersona, Persona
-
-# expose model providers on the package root
-from .providers import (
-    AI21Provider,
-    BaseProvider,
-    GPT4AllProvider,
-    HfHubProvider,
-    QianfanProvider,
-    TogetherAIProvider,
-)
+_dynamic_imports = {
+    # expose embedding model providers on the package root
+    "embedding_providers": [
+        "BaseEmbeddingsProvider",
+        "GPT4AllEmbeddingsProvider",
+        "HfHubEmbeddingsProvider",
+        "QianfanEmbeddingsEndpointProvider",
+    ],
+    "exception": ["store_exception"],
+    "magics": ["AiMagics"],
+    # expose JupyternautPersona on the package root
+    # required by `jupyter-ai`.
+    "models.persona": ["JupyternautPersona", "Persona"],
+    # expose model providers on the package root
+    "providers": [
+        "AI21Provider",
+        "BaseProvider",
+        "GPT4AllProvider",
+        "HfHubProvider",
+        "QianfanProvider",
+        "TogetherAIProvider",
+    ],
+}
+
+_dynamic_imports_map = {
+    import_name: module
+    for module, imports in _dynamic_imports.items()
+    for import_name in imports
+}
+
+
+def __getattr__(attr_name: str) -> object:
+    module_name = _dynamic_imports_map.get(attr_name)
+    result = import_attr(attr_name, module_name, __spec__.parent)
+    globals()[attr_name] = result
+    return result
 
 
 def load_ipython_extension(ipython):
-    ipython.register_magics(AiMagics)
-    ipython.set_custom_exc((BaseException,), store_exception)
+    ipython.register_magics(__getattr__("AiMagics"))
+    ipython.set_custom_exc((BaseException,), __getattr__("store_exception"))
 
 
 def unload_ipython_extension(ipython):
     ipython.set_custom_exc((BaseException,), ipython.CustomTB)
+
+
+# required to preserve backward compatibiliy with `from jupyter_ai_magics import *`
+__all__ = [
+    "__version__",
+    "load_ipython_extension",
+    "unload_ipython_extension",
+    "BaseEmbeddingsProvider",
+    "GPT4AllEmbeddingsProvider",
+    "HfHubEmbeddingsProvider",
+    "QianfanEmbeddingsEndpointProvider",
+    "store_exception",
+    "AiMagics",
+    "JupyternautPersona",
+    "Persona",
+    "AI21Provider",
+    "BaseProvider",
+    "GPT4AllProvider",
+    "HfHubProvider",
+    "QianfanProvider",
+    "TogetherAIProvider",
+]
+assert set(_dynamic_imports_map.keys()) - set(__all__) == set()
```

Comment:
This static assertion is great, but it won't run until the module is imported. Could we test this to guarantee that CI will block release if this assertion excepts?

```
def test_top_level_module():
  from jupyter_ai_magics import * as exports
  assert hasattr(exports, 'BaseProvider')
  ... # etc